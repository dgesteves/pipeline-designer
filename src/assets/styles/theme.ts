import "styled-components";
import { DefaultTheme } from "styled-components";

const theme: DefaultTheme = {
  colors: {
    white: "#ffffff",
    background: "#1D1E22",
    violet: "#c43ad6",
    turquoise: "#30CFD0",
  },
  sizes: {
    size_1: "0.625rem",
    size_2: "0.75rem",
    size_3: "0.875rem",
    size_4: "1rem",
    size_5: "1.125rem",
    size_6: "1.25rem",
    size_7: "1.5rem",
    size_8: "1.875rem",
    size_9: "2rem",
  },
  gradients: {
    textGradient: "linear-gradient(to right, #30CFD0, #c43ad6)",
  },
  shadows: {
    textShadow:
      "0 0 80px rgb(192 219 255 / 35%), 0 0 32px rgb(65 120 255 / 24%)",
  },
};

export default theme;
