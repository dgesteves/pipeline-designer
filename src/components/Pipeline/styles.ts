import styled from "styled-components";

export const PipelineSection = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  flex: 1;
`;

export const Node = styled.div`
  display: flex;
  justify-content: center;
  gap: ${({ theme }) => theme.sizes.size_4};
  align-items: center;
  width: ${({ theme }) => `calc(${theme.sizes.size_9}*6)`};
  height: ${({ theme }) => `calc(${theme.sizes.size_9}*2)`};
  color: ${({ theme }) => theme.colors.white};
  border: 2px double transparent;
  background-image: linear-gradient(
      ${({ theme }) => theme.colors.background},
      ${({ theme }) => theme.colors.background}
    ),
    radial-gradient(
      circle at left top,
      ${({ theme }) => theme.colors.turquoise},
      ${({ theme }) => theme.colors.violet}
    );
  background-origin: border-box;
  background-clip: padding-box, border-box;
  padding: ${({ theme }) => theme.sizes.size_4};
  border-radius: 8px;
  color: ${({ theme }) => theme.colors.white};
  text-transform: uppercase;

  svg {
    cursor: pointer;
  }
`;

export const Edge = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  font-size: ${({ theme }) => theme.sizes.size_6};
`;

export const Line = styled.div`
  height: ${({ theme }) => theme.sizes.size_1};
  width: 2px;
  background-color: ${({ theme }) => theme.colors.white};
`;
