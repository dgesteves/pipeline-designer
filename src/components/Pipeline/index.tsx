import { Edge, Line, Node, PipelineSection } from "./styles";
import { DragEvent, Fragment, useCallback } from "react";
import { v4 as uuid } from "uuid";
import Add from "../../assets/icons/Add";
import Remove from "../../assets/icons/Remove";
import { INode, TName } from "../../services/types";
import useFetchNodes from "../../hooks/useFetchNodes";

export default function Pipeline() {
  const { nodes, setNodes, isLoading, error } = useFetchNodes();

  const onRemove = useCallback(
    (id: string) => {
      setNodes((nodes) => nodes.filter((node) => node.id !== id));
    },
    [setNodes]
  );

  const onDrop = useCallback(
    (e: DragEvent<SVGSVGElement>, index: number) => {
      e.preventDefault();
      const id = uuid();
      const name = e.dataTransfer.getData("name") as TName;
      setNodes((nodes: INode[]) => {
        const newNodes = [...nodes];
        newNodes.splice(index + 1, 0, { id, name });
        return newNodes;
      });
    },
    [setNodes]
  );

  const onDragOver = useCallback((e: DragEvent<SVGSVGElement>) => {
    e.preventDefault();
  }, []);

  if (isLoading) {
    return <p>Loading...</p>;
  }

  if (error) {
    return <p>Error: {error} </p>;
  }

  return (
    <PipelineSection data-testid="pipeline-section-test">
      {nodes?.map(({ name, id }, i) => (
        <Fragment key={id}>
          <Node data-testid="node-test">
            <p>{name}</p>
            {name !== "query" && (
              <Remove
                data-testid="remove-node-test"
                onClick={() => onRemove(id)}
              />
            )}
          </Node>
          <Edge>
            <Line />
            <Add
              data-testid="add-node-test"
              onDrop={(e) => onDrop(e, i)}
              onDragOver={onDragOver}
            />
            {id !== nodes?.at(-1)?.id && <Line />}
          </Edge>
        </Fragment>
      ))}
    </PipelineSection>
  );
}
