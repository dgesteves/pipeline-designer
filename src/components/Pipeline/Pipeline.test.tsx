import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import Pipeline from "./index";
import { ThemeProvider } from "styled-components";
import theme from "../../assets/styles/theme";
import GlobalStyles from "../../assets/styles/globalStyles";

describe("Pipeline component", () => {
  it("should render correctly", () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <Pipeline />
      </ThemeProvider>
    );

    expect(container).toBeInTheDocument();
  });

  it("should render initial state", () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <Pipeline />
      </ThemeProvider>
    );

    expect(container.textContent).toBe("Loading...");
  });

  it("should match previous snapshot", async () => {
    render(
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <Pipeline />
      </ThemeProvider>
    );

    await waitFor(() => {
      screen.getByTestId("pipeline-section-test");
    });
    expect(screen.getByTestId("pipeline-section-test")).toMatchSnapshot();
  });

  it("should remove node correctly when clicked", async () => {
    render(
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <Pipeline />
      </ThemeProvider>
    );

    await waitFor(() => {
      screen.getByTestId("pipeline-section-test");
    });

    expect(screen.getAllByTestId("node-test")).toHaveLength(3);

    fireEvent.click(screen.getAllByTestId("remove-node-test").at(0)!);

    expect(screen.getAllByTestId("node-test")).toHaveLength(2);

    fireEvent.click(screen.getAllByTestId("remove-node-test").at(0)!);

    expect(screen.getAllByTestId("node-test")).toHaveLength(1);
  });

  it("should have query as first node on pipeline and can not be removed", async () => {
    render(
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <Pipeline />
      </ThemeProvider>
    );

    await waitFor(() => {
      screen.getByTestId("pipeline-section-test");
    });

    const queryNode = screen.getAllByTestId("node-test").at(0);

    expect(queryNode?.textContent).toBe("query");

    fireEvent.click(screen.getAllByTestId("remove-node-test").at(0)!);

    expect(screen.getAllByTestId("node-test")).toHaveLength(2);

    fireEvent.click(screen.getAllByTestId("remove-node-test").at(0)!);

    expect(screen.getAllByTestId("node-test")).toHaveLength(1);

    expect(queryNode).toBeInTheDocument();
  });

  it("should add node correctly when dropped", async () => {
    render(
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <Pipeline />
      </ThemeProvider>
    );

    await waitFor(async () => {
      screen.getByTestId("pipeline-section-test");
    });

    expect(screen.getAllByTestId("node-test")).toHaveLength(3);

    const addIcon = screen.getAllByTestId("add-node-test").at(0);
    const mockData = { getData: jest.fn() };

    fireEvent.drop(addIcon!, {
      dataTransfer: mockData,
    });

    expect(screen.getAllByTestId("node-test")).toHaveLength(4);

    fireEvent.drop(addIcon!, {
      dataTransfer: mockData,
    });

    expect(screen.getAllByTestId("node-test")).toHaveLength(5);
  });

  it("should get data onDrop", async () => {
    render(
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <Pipeline />
      </ThemeProvider>
    );

    await waitFor(() => {
      screen.getByTestId("pipeline-section-test");
    });

    const addIcon = screen.getAllByTestId("add-node-test").at(0);
    const mockData = { getData: jest.fn() };

    fireEvent.drop(addIcon!, {
      dataTransfer: mockData,
    });

    expect(mockData.getData).toHaveBeenCalled();
  });
});
