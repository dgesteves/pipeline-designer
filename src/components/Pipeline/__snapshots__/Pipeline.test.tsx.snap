// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pipeline component should match previous snapshot 1`] = `
<div
  class="sc-bczRLJ HfAct"
  data-testid="pipeline-section-test"
>
  <div
    class="sc-gsnTZi bJBqjV"
    data-testid="node-test"
  >
    <p>
      query
    </p>
  </div>
  <div
    class="sc-dkzDqf eQXDNY"
  >
    <div
      class="sc-hKMtZM bZnjne"
    />
    <svg
      data-testid="add-node-test"
      height="1em"
      viewBox="0 0 32 32"
      width="1em"
      xml:space="preserve"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16 4c6.6 0 12 5.4 12 12s-5.4 12-12 12S4 22.6 4 16 9.4 4 16 4m0-2C8.3 2 2 8.3 2 16s6.3 14 14 14 14-6.3 14-14S23.7 2 16 2z"
        style="fill: currentColor;"
      />
      <path
        d="M24 15h-7V8h-2v7H8v2h7v7h2v-7h7z"
        style="fill: currentColor;"
      />
      <path
        d="M0 0h32v32H0z"
        style="fill: none;"
      />
    </svg>
    <div
      class="sc-hKMtZM bZnjne"
    />
  </div>
  <div
    class="sc-gsnTZi bJBqjV"
    data-testid="node-test"
  >
    <p>
      retriever
    </p>
    <svg
      data-testid="remove-node-test"
      height="1em"
      viewBox="0 0 32 32"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 12h2v12h-2zM18 12h2v12h-2z"
        style="fill: currentColor;"
      />
      <path
        d="M4 6v2h2v20a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V8h2V6Zm4 22V8h16v20ZM12 2h8v2h-8z"
        style="fill: currentColor;"
      />
      <path
        d="M0 0h32v32H0z"
        data-name="<Transparent Rectangle>"
        style="fill: none;"
      />
    </svg>
  </div>
  <div
    class="sc-dkzDqf eQXDNY"
  >
    <div
      class="sc-hKMtZM bZnjne"
    />
    <svg
      data-testid="add-node-test"
      height="1em"
      viewBox="0 0 32 32"
      width="1em"
      xml:space="preserve"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16 4c6.6 0 12 5.4 12 12s-5.4 12-12 12S4 22.6 4 16 9.4 4 16 4m0-2C8.3 2 2 8.3 2 16s6.3 14 14 14 14-6.3 14-14S23.7 2 16 2z"
        style="fill: currentColor;"
      />
      <path
        d="M24 15h-7V8h-2v7H8v2h7v7h2v-7h7z"
        style="fill: currentColor;"
      />
      <path
        d="M0 0h32v32H0z"
        style="fill: none;"
      />
    </svg>
    <div
      class="sc-hKMtZM bZnjne"
    />
  </div>
  <div
    class="sc-gsnTZi bJBqjV"
    data-testid="node-test"
  >
    <p>
      reader
    </p>
    <svg
      data-testid="remove-node-test"
      height="1em"
      viewBox="0 0 32 32"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 12h2v12h-2zM18 12h2v12h-2z"
        style="fill: currentColor;"
      />
      <path
        d="M4 6v2h2v20a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V8h2V6Zm4 22V8h16v20ZM12 2h8v2h-8z"
        style="fill: currentColor;"
      />
      <path
        d="M0 0h32v32H0z"
        data-name="<Transparent Rectangle>"
        style="fill: none;"
      />
    </svg>
  </div>
  <div
    class="sc-dkzDqf eQXDNY"
  >
    <div
      class="sc-hKMtZM bZnjne"
    />
    <svg
      data-testid="add-node-test"
      height="1em"
      viewBox="0 0 32 32"
      width="1em"
      xml:space="preserve"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16 4c6.6 0 12 5.4 12 12s-5.4 12-12 12S4 22.6 4 16 9.4 4 16 4m0-2C8.3 2 2 8.3 2 16s6.3 14 14 14 14-6.3 14-14S23.7 2 16 2z"
        style="fill: currentColor;"
      />
      <path
        d="M24 15h-7V8h-2v7H8v2h7v7h2v-7h7z"
        style="fill: currentColor;"
      />
      <path
        d="M0 0h32v32H0z"
        style="fill: none;"
      />
    </svg>
  </div>
</div>
`;
